### PR TITLE
DistMatrix: Fix use of DistMatrix

### PR DESCRIPTION
**Pull request by ales-erjavec**

### CODE REVIEW

**Comment by kernc** [comment]
Sorry. I, too, hate NumPy breaking the valid-object-evaluates-to-True-in-boolean-contexts contract. But sadly, there's no alternative. :disappointed: 

While you're at it, owmds.py:465 has `self._effective_matrix = Orange.distance.Euclidean(preprocessed_data)` which is on 492 ... used as `self._effective_matrix.X`, which warns of deprecation.

File: Orange/clustering/hierarchical.py [hunk 1]
```diff
@@ -355,13 +355,14 @@ def optimal_leaf_ordering(tree, distances, progress_callback=None):
...
-        DistMatrix that was used to compute the clustering.
+    :param numpy.ndarray distances:
+        A (N, N) numpy.ndarray of distances that were used to compute
+        the clustering.
     :param function progress_callback:
         Function used to report on progress.
 
     """
-    distances = distances.X
+    distances = numpy.asarray(distances)
```
Comment (lanzagar) [inline Orange/clustering/hierarchical.py]: Why is this line needed?
Since distances already are a numpy.ndarray it does nothing, right?

File: Orange/widgets/unsupervised/owhierarchicalclustering.py [hunk 2]
```diff
@@ -944,8 +944,8 @@ def _update(self):
 
         if distances is not None:
             # Convert to flat upper triangular distances
-            i, j = numpy.triu_indices(distances.X.shape[0], k=1)
-            distances = distances.X[i, j]
+            i, j = numpy.triu_indices(distances.shape[0], k=1)
+            distances = numpy.asarray(distances[i, j])
```
Comment (lanzagar) [inline Orange/widgets/unsupervised/owhierarchicalclustering.py]: Also here...
(we called distances.shape above so distances really shouldn't be something other than a np.ndarray)

File: Orange/clustering/hierarchical.py (hunk 1 above)
Comment (ales-erjavec) [inline Orange/clustering/hierarchical.py]: `distances` might be a ndarray  'subclass' (for instance a numpy.matrix). If so then asarray creates an array that is a view of it's contents.

File: Orange/widgets/unsupervised/owhierarchicalclustering.py (hunk 2 above)
Comment (ales-erjavec) [inline Orange/widgets/unsupervised/owhierarchicalclustering.py]: Here distances will always be a DistMatrix.  But there is something unnerving about a 'symmetric matrix' with shape (N, ).

File: Orange/clustering/hierarchical.py (hunk 1 above)
Comment (lanzagar) [inline Orange/clustering/hierarchical.py]: Before the change it was a DistMatrix. Now, let's say we are more general, but even if an exotic subclass is provided, it only needs to support indexing (line 370), then it is again converted to an (contiguous) array.

But the added check does not hurt since it does not do much in normal circumstances.

File: Orange/widgets/unsupervised/owhierarchicalclustering.py (hunk 2 above)
Comment (kernc) [inline Orange/widgets/unsupervised/owhierarchicalclustering.py]: It has a shape (N, N). And I foresaw the frequent need of upper triangular, so `DistMatrix.flat()` is also provided. `triu()` might be better, now that I think about it.

Comment (ales-erjavec) [inline Orange/widgets/unsupervised/owhierarchicalclustering.py]:
```python
D = Orange.misc.DistMatrix(numpy.eye(4))
i, j = numpy.triu_indices(D.shape[0], 1)
print("Shape:", D[i, j].shape)
print(repr(D[i, j]))
```
outputs
```
Shape: (6,)
DistMatrix([ 0.,  0.,  0.,  0.,  0.,  0.])
```

Comment (kernc) [inline Orange/widgets/unsupervised/owhierarchicalclustering.py]: Sure, but that's because `D[i, j]` is no longer a symmetric matrix; it's a 1D array constucted anew from `i` and `j` indices. It's the same with
```py
>>> A = np.arange(D.size).reshape(D.shape)
>>> A[i, j]
array([ 1,  2,  3,  6,  7, 11])
```
There is no problem, right?
